Initial development

lette1394 opened the pull request:
전체 root directory에서 
각 sub project를 생성하는 방식으로 했고, 빌드 도구는 gradle을 사용함. 

다른 사람의 코드를 자주 보게끔 유도하는 의미에서, 전체 git을 최상위에 하나만 두었다. 
사실 git을 sub directory 별로 할 수 있는지 모르겠다. 아마 할 수 있겠지? 근데 그러면 다른 사람 코드 보기 너무 귀찮아지니까 그냥 최상위에 git을 하나만 두었다.

또한 현재까지 구현은 command를 parsing하는 것까지 했다. 
Commit 이름을 붙이는 전략을 정해야 할 것 같다. (e. g. [<issue-number>] <title...>

lette1394 committed:
Update .gitignore

lette1394 committed:
Add sub module of jae-eun-na-1

lette1394 committed:
Add dependency for test

junit 5, assertj

lette1394 committed:
Update READMD.md add 학습목표

lette1394 committed:
Add command line test for init command

프로그램 시작하고 새로운 게임을 초기화

lette1394 committed:
Add implements for MainTest

테스트 통과를 위해 asString()을 추가

lette1394 committed:
Add test for init command when there is already stared game

lette1394 committed:
Remove init test, Add test for show, unknown command

현재 단계에서 init command의 다른 테스트를 추가하는 것은, 적절하지 않다고 여겨짐. init command의 추가 적인 기능이 다른 로직에 의존하고 있기 때문.
대신에 다른 show, unknown command에 대한 기본적인 동작을 테스트 할 코드를 작성함.

lette1394 committed:
Add implements for show, unknown command

현재까지는 단순히 command를 구분하는 정도로만 구현

lette1394 committed:
Extract method in MainTest to mock standard input, output stream

ˆ

lette1394 committed:
Rename Main class

lette1394 committed:
Extract logic of BowlingScoreBoard and delegate to CommandProcessor

볼링 명세에 적힌 커맨드를 처리하는 방법을 아는 객체를 생성하고, 입력을 위임하여 처리

lette1394 committed:
Add Command enum to remove Strings in switch-case statement

팩터리 메서드 of에 대해 : 일치하지 않는 입력이 발생해서 예외가 났을 때,
1. fallback 콜백을 받아서 Command enum이 처리하는 방법
2. 예외를 먹어버리고 UNKNOWN command를 반환하는 방법

두 가지 방법을 생각할 수 있었다. 처음엔 1번으로 코딩했다가 CommandProcessor 에 있는 switch-case 문에서 일관성이 떨어지는 것 같아, 다시 2번의 형태로 변경했다.

lette1394 committed:
Add Input, Output interface and Add implements of that in command line

CommandProcessor와 Input source와의 결합도를 낮추려는 것이 목적이다.
기존에 InputStream으로 받아서 어느정도 목적을 이뤘었지만, 자바 기본 인터페이스 이므로 한 번 감싸서 유연성을 주고 싶었다.

그 결과가 Input, Output 인터페이스이며, 현재 스펙상으로는 파일입출력 혹은 네트워크 입출력이 없으므로 Command line에 대한 구현체만 구현하였다.
각 구현체는 다음 Command가 있는지, 있다면 무엇인지, 현재 Command에 대한 Argument 등을 가져오는 방법들을 안다.

lette1394 committed:
Apply Output interface to CommandProcessor

앞서 구현한 CommandLineOutput을 CommandProcessor에 적용했다.
이로서 Processor는
1. 자신이 처리해야 할 작업이 어디서 오는지,
2. 자신이 처리한 결과가 어디로 가는지
모르는 상태가 되며, 이는 원래부터 알 필요가 없으므로 바람직하다.

현재 Processor는 두 가지 일을 하는데,
1. Command에 따라서 작업을 분리하는 일 (switch-case를 통해)
2. Command에 맞는 작업을 실행하는 일 (case안의 중괄호)
이는 분리해야 한다.

lette1394 committed:
Add comments for implements how divide input and parsing

주석 참고. 구현 후에 꼭 지우자.

eminem54 commented:
이슈에 보니까 그래들사용하시던데 어디에 사용하시나요? 
그리고 깃을 최상위에 하나 둔다는게 리포하나를 가지고 작업하는거니까 작업하는사람리 다른사람의 소스코드까지 가지고있어야한다는건가요? 
그건 그거 나름대로 관리하기가 어려웁지않나요?

lette1394 commented:
1. Gradle은 빌드 스크립트로 사용합니다. 저도 Gradle 잘 몰라서 task 이런건 쓸 줄 모르고요, 공부하셔서 공유해주시면 감사하겠습니다 (감사)
2. 각자 브랜치를 파서 작업하는 것이 권장되고, 그렇게 하지 않더라도 디렉토리 별로 파일이 구분되므로 괜찮지 않을까...생각했습니다.

lette1394 committed:
Add CommandContext class, CommandHandler interface, etc.

CommandArgument 안의 값은 무엇이 들어올지 모르기 때문에 class로 선언했다..
CommandOption 의 값은 미리 정해둬야 하기 때문에 enum으로 선언했다.
CommandHandler의 입력과 출력이 중요한데, 두 가지 후보가 있다.

1. 입력 : CommandContext, 출력 : String
handler에게 필요한 입력(Input)은 모두 CommandContext에게 있다.
커맨드가 실행되고, 유저에게 보여지는 값은 어짜피 String 형태일 것이라 예상할 때 가능하다.
그러나 이렇게 선언한다면, 내부 상태를 유지하는 값이 변경 될 것이 보이지 않기 때문에 테스트가 어렵고 순수함수가 아닐 가능성이 높다.

2. 입력 : CommandContext, Output, 출력: void
처리된 결과를 handler 안에서 처리하느냐, 밖에서 처리하느냐 문제지, 이 역시 1번과 같다.

3. 입력 : CommandContext, BowlingContext, 출력 : Result(String) **(낙찰)**
현재로서는 handler의 출력이 유저에게 보이는 CLI 인터페이스에만 필요하다고 가정한다.
추후에 Json이나 Xml등으로 표현하기 위해서 CommandResult 같은 반환 인터페이스를 검토해볼 수 있겠다.
또한 모든 커맨드에 대해서 BowlingContext를 기반으로 움직이게 한다.
커맨드를 실행하기 위한 의존적인 모든 상태 정보는 BowlingContext에 들어있으며, 여기에는 현재 Frame 정보, User 정보, Score 정보 등이 들어가 있다.
각 커맨드별 필요한 정보들이 다르기 때문에 공통 정보를 제외한 나머지 정보들은 필드로 Map<Class, DataClass> 형태를 고려해 볼 수 있다.

lette1394 committed:
Rename BowlingContext -> BowlingState

이름이 ...Context하면서 Context라는 이름을 너무 남발하는 것 같다.
CommandContext를 바꾸려 했는데, Command enum의 이름을 바꾸기가 너무 어렵다.
사실 `$ show winner -f` 이런식으로 입력하면 Command라고 하면 `show`만 얘기하는 것 보다, 저기 있는 모든 부분을 다 말하는 게 된다.
근데 맨 앞의 `show`를 어떻게 표현할지 어렵다.

그래서 BowlingContext를 State로 변경했다.
애초에 state 패키지였으니, 더 적절한 이름이다.

또한 BowlingState에 기대하는 기능을 따져보면, 한 번 쓰고 버리는 것이 아닌 계속 정보를 들고 있는 전역상태처럼 행동하므로 적절해졌다.

lette1394 committed:
Add comments for future implements of removing while in CommandProcessor

lette1394 committed:
Update build config

lette1394 committed:
Extract standard input output mock method to test.util

lette1394 committed:
Add CommandContextParserTest

이름 최종 낙찰 : CommandContextParser
이 클래스는 Input 에서 CommandContext를 생성하는 방법을 안다.

그런데 CommandContextParser는 필연적으로 Input 인터페이스에 의존적이므로,
Input -> Parser로 넘어오는 객체가 현재는 String이다. 그렇다는 말은 문자열 기반이라는건데...
나중에 HTTP call이나 JSON 형태로 넘어오는 것도 String line으로 말아줘야 한다는 말이다.

일단은 이렇게 진행하도록 하자.

lette1394 committed:
Add implements for CommandContextParserTest

lette1394 committed:
Add factory method to Output

Input과 대칭을 맞추기 위해 Output에도 팩터리메서드 도입.
또한 범용성을 위해서 OutputStream을 인자로 받음
